### PR TITLE
118 data schema

### PR DESCRIPTION
**Pull request by RoddyMacrae**
## Pull Request submission 

*Insert detailed bullet points about your changes here!*
- This should in theory only be needed once
- xlsx file had to be downloaded and run locally as connecting to Sharepoint directly proved too difficult and not needed in the scope of this ticket
- data_schema.py can be found in src/data_validation
- DataSchema.toml is located in config

*Insert any instructions to help the reviewer, e.g. "install new requirements from `environment.yml`"*
- Download the BERD xlsx file locally along with the data_schema.py
- You may want to change where DataSchema.toml is saved to. It is currently saved to the same directory.

*Let the reviewer know what data files are needed (and if applicable, where they are to be found)
- Download the BERD Data Dictionary locally for use: https://officenationalstatistics.sharepoint.com/:x:/r/sites/ESCSEPDS/_layouts/15/doc2.aspx?sourcedoc=%7Be1685b01-3318-451e-b413-4f551ad143bf%7D&action=edit&activeCell=%27Sheet1%27!B10&wdinitialsession=306d0a09-707b-4b33-8cc0-0f388f8ccf54&wdrldsc=2&wdrldc=1&wdrldr=OpenACopyInExcel%2CAccessTokenExpiredWarning%2CDeploym

#### Closes or fixes

* Detail the ticket(s) you are closing with this PR
Closes #

#### Code

- [x] **Code runs** The code runs on my machine and/or CDSW
- [ ] **Conflicts resolved** There are no conflicts (I have performed a rebase if necessary)
- [x] **Requirements** My/our code functions according to the requirements of the ticket
- [x] **Dependencies** I have updated the environment yaml so it includes any new libraries I have used
- [ ] **Configuration file updated** any high level parameters that the user may interact with have been put into the config file (and imported to the script) 
- [x] **Clean Code**
    - [x] Code is as [PEP 8]([url](https://peps.python.org/pep-0008/)) compliant as I can humanly make it
    - [x] Code passess flake8 linting check
    - [x] Code adheres to [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)
- [x] **Type hints** All new functions have [type hints ](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html)

#### Documentation

Any new code includes all the following forms of documentation:

- [x] **Function Documentation** Docstrings within the function(s')/methods have been created
    - [x] Includes `Args` and `returns` for all major functions 
    - [x] The docstring details data types
- [ ] **Updated Documentation**: User and/or developer working doc has been updated

#### Data
- [ ] All data needed to run this script is available in Dev/Test
- [x] All data is excluded from this pull request
- [x] Secrets checker pre-commit passes

#### Testing
- [ ] **Unit tests** Unit tests have been created and are passing _or a new ticket to create tests has been created_

---

# Peer Review Section

- [ ] All requirements install from (updated) `environment.yaml`
- [ ] Documentation has been created and is clear - **check the working document**
- [ ] Doctrings ([Google format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html)) have been created and accurately describe the function's functionality
- [ ] Unit tests pass, or if not present _a new ticket to create tests has been created_
- [ ] **Code runs** The code runs on reviewer's machine and/or CDSW

#### Final approval (post-review)

The author has responded to my review and made changes to my satisfaction.
- [ ] **I recommend merging this request.**

---

### Review comments

*Insert detailed comments here!*

These might include, but not exclusively:

- bugs that need fixing (does it work as expected? and does it work with other code
  that it is likely to interact with?)
- alternative methods (could it be written more efficiently or with more clarity?)
- documentation improvements (does the documentation reflect how the code actually works?)
- additional tests that should be implemented (do the tests effectively assure that it
  works correctly?)
- code style improvements (could the code be written more clearly?)
- Do the changes represent a change in functionality so the version number should increase? Start a discussion if so.
- As a review you can generates the same outputs from running the code

Your suggestions should be tailored to the code that you are reviewing.
Be critical and clear, but not mean. Ask questions and set actions.

### CODE REVIEW

**Comment by gallmand** [comment]
There is an old version of data_schema.py in the higher level src folder that I think is now redundant.
Also there are trailing spaces in the description strings in the toml file, we could do to remove these.

**Comment by RoddyMacrae** [comment]
All recommended changes have been completed. Ready for re-review/merge.

**Comment by gallmand** [comment]
On lines 139 to 142 of data_schema.py - it is worth giving the test variables more descriptive names.

File: src/data_validation/data_schema.py [hunk 1]
```diff
@@ -136,7 +137,7 @@ def create_toml(pdDict: dict) -> IO[str]:
...
 
 
-test = read_DAP_csv("/ons/rdbe_dev/data_dictionary_berd.csv")
-test2 = convert_dataFrame(test)
-test3 = reformat_tomlDict(test2)
-test4 = create_toml(test3)
+csv_dataframe = read_DAP_csv("/ons/rdbe_dev/data_dictionary_berd.csv")
+csv_dataDict = convert_dataFrame(csv_dataframe)
+reformated_Dict = reformat_tomlDict(csv_dataDict)
+tomlFile = create_toml(reformated_Dict)
```
Comment (jwestw) [inline src/data_validation/data_schema.py:143]: Please check point 6 in this document on Python naming conventions: https://visualgit.readthedocs.io/en/latest/pages/naming_convention.html

Comment (jwestw) [reply src/data_validation/data_schema.py:143]: These names are better, but still generic. "csv_dataframe" could refer to any dataframe that's come from a csv. Think about what the data contains. Also, to save space we often use `_df` as a suffix.

Comment (RoddyMacrae) [reply src/data_validation/data_schema.py]: Thank you for the link. I have gone over it and bookmarked it for future reference! 

I have also updated the naming on lines 140 - 143 to be more descriptive and match naming conventions. Hopefully they are better.

File: config/Data_Schema.toml [hunk 2]
```diff
@@ -0,0 +1,469 @@
+[snapshot_id]
+Description = "nan"
+Deduced_Data_Type = "nan"
+Nullable = "nan"
```
Comment (jwestw) [inline config/Data_Schema.toml:4]: Why is nullable always "nan" 0.0 or 1.0? Is this equal to N/A, True and False?

File: config/Data_Schema.toml [hunk 3]
```diff
@@ -0,0 +1,469 @@
...
+Nullable = 0.0
+Current_Data_Type = "int"
+Length = "nan"
+Min_values = "nan"
+Max_values = "nan"
+Possible_Categorical_Values = [0]
+
+[currency]
+Description = "nan"
+Deduced_Data_Type = "str"
```
Comment (jwestw) [inline config/Data_Schema.toml:273]: This looks like it should be categorical data

File: config/Data_Schema.toml [hunk 4]
```diff
@@ -0,0 +1,469 @@
...
+Nullable = 0.0
+Current_Data_Type = "str"
+Length = "nan"
+Min_values = "nan"
+Max_values = "nan"
+Possible_Categorical_Values = ["VATREF"]
+
+[payereference]
+Description = "nan"
+Deduced_Data_Type = "str"
```
Comment (jwestw) [inline config/Data_Schema.toml:293]: Looks like categorical data to me

File: config/Data_Schema.toml [hunk 5]
```diff
@@ -0,0 +1,469 @@
...
+Nullable = 0.0
+Current_Data_Type = "int"
+Length = 2
+Min_values = "nan"
+Max_values = "nan"
+Possible_Categorical_Values = [20, 21]
+
+[status]
+Description = "nan"
+Deduced_Data_Type = "str"
```
Comment (jwestw) [inline config/Data_Schema.toml:53]: I think these are categories

File: config/Data_Schema.toml [hunk 6]
```diff
@@ -0,0 +1,469 @@
...
+Nullable = 1.0
+Current_Data_Type = ["None","str"]
+Length = 0
+Min_values = "nan"
+Max_values = "nan"
+Possible_Categorical_Values = ["nan"]
+
+[lockeddate]
+Description = "All None type"
+Deduced_Data_Type = "?"
```
Comment (jwestw) [inline config/Data_Schema.toml:93]: pandas.NA maybe?

File: config/Data_Schema.toml [hunk 7]
```diff
@@ -0,0 +1,469 @@
...
+Nullable = 0.0
+Current_Data_Type = "str"
+Length = "nan"
+Min_values = "nan"
+Max_values = "nan"
+Possible_Categorical_Values = ["0001", "0006"]
+
+[checkletter]
+Description = "nan"
+Deduced_Data_Type = "str"
```
Comment (jwestw) [inline config/Data_Schema.toml:113]: Category

File: config/Data_Schema.toml [hunk 8]
```diff
@@ -0,0 +1,469 @@
...
+Nullable = 0.0
+Current_Data_Type = "str"
+Length = "nan"
+Min_values = "nan"
+Max_values = "nan"
+Possible_Categorical_Values = ["S", "E"]
+
+[vatreference]
+Description = "nan"
+Deduced_Data_Type = "str"
```
Comment (jwestw) [inline config/Data_Schema.toml:283]: Category

File: config/Data_Schema.toml (hunk 2 above)
Comment (RoddyMacrae) [reply config/Data_Schema.toml:4]: Yes it is. I have changed it so it always read True or False.

File: config/Data_Schema.toml (hunk 3 above)
Comment (RoddyMacrae) [reply config/Data_Schema.toml]: Yep, exactly right. Amended this and all others now.

File: config/Data_Schema.toml [hunk 9]
```diff
@@ -0,0 +1,469 @@
...
+Max_values = "nan"
+Possible_Categorical_Values = ["nan"]
+
+[reference]
+Description = "nan"
+Deduced_Data_Type = "int"
+Nullable = 0.0
+Current_Data_Type = "str"
+Length = ">=1"
+Min_values = 11001603625.0
```
[the remark this answers is not in the document]
Comment (RoddyMacrae) [reply config/Data_Schema.toml]: Yeah, good spot. Thanks! Fixed this and all others throughout toml file.

File: config/Data_Schema.toml (hunk 6 above)
Comment (RoddyMacrae) [reply config/Data_Schema.toml]: Yeah, I think so. Certainly, until further data/something else comes through to suggest otherwise!

**Comment by jwestw** [comment]
@ali-i3 

I am running @RoddyMacrae 's code here and getting this error related to `config_df` in the run log

image

The creation of `config_df` is inside `retrieve_configs()` but that function doesn't run at any point. Can you help me take a look at that next week?

**Comment by ali-i3** [comment]
> @ali-i3
> 
> I am running @RoddyMacrae 's code here and getting this error related to `config_df` in the run log
> 
> image
> 
> The creation of `config_df` is inside `retrieve_configs()` but that function doesn't run at any point. Can you help me take a look at that next week?

`runlog_obj.retrieve_configs()` is missing from `src/main.py` which captures the latest configs for the run. I'll add it now.

**Review by jwestw — APPROVED** [review]
All working now.

**Pull request merged**